modules/modelSetup/mixin/ModelSetupDiffusionLossMixin.py

```diff
@@ -156,21 +157,52 @@ def __min_snr_weight(
             gamma: float,
             v_prediction: bool,
             device: torch.device
-    ):
+    ) -> Tensor:
         if self.__coefficients:
             all_snr = (self.__coefficients.sqrt_alphas_cumprod /
                        self.__coefficients.sqrt_one_minus_alphas_cumprod) ** 2
             all_snr.to(device)
             snr = all_snr[timesteps]
         else:
             alphas_cumprod = self.__alphas_cumprod_fun(timesteps, 1)
-            snr = ((alphas_cumprod ** 0.5) / (1.0 - alphas_cumprod) ** 0.5) ** 2
+            snr = alphas_cumprod / (1.0 - alphas_cumprod)
         min_snr_gamma = torch.minimum(snr, torch.full_like(snr, gamma))
         # Denominator of the snr_weight increased by 1 if v-prediction is being used.
         if v_prediction:
             snr += 1
-        snr_weight = torch.div(min_snr_gamma, snr).float().to(device)
+        snr_weight = (min_snr_gamma / snr).to(device)
         return snr_weight
+    
+    def __debiased_estimation_weight(
+        self,
+        timesteps: Tensor,
+        device: torch.device
+    ) -> Tensor:
+        if self.__coefficients:
+            all_snr = (self.__coefficients.sqrt_alphas_cumprod /
+                       self.__coefficients.sqrt_one_minus_alphas_cumprod) ** 2
+            all_snr.to(device)
+            snr = all_snr[timesteps]
+        else:
+            alphas_cumprod = self.__alphas_cumprod_fun(timesteps, 1)
+            snr = alphas_cumprod / (1.0 - alphas_cumprod)
+        weight = snr
```

Comment:
Shouldn't this be `weight = 1/snr` instead?
From the paper (equation 11):

![grafik](https://github.com/Nerogar/OneTrainer/assets/3390934/ec9d6677-74f5-44f1-af6a-ee1881051bf5)

`weight = snr` looks like it would work for x_0 prediction instead of eps prediction.

Reply:
I added a new method `__snr` to calculate the SNR, it should by my intuition handle all cases, but maybe @mx should take a look.